design/upgrade.md

```diff
@@ -0,0 +1,165 @@
+
+## **Goal**
+In order for software upgrade support in Rook to be considered successful, the goals listed below should be met.
+Note that these goals are for a long term vision and are not all necessarily deliverable within the v0.6 release time frame.
+* **Automatic:** When a new version of Rook is released, a live cluster should be able to update all its components to the new version without user intervention.
+* **No downtime:** During an upgrade window, there should be **zero** downtime of cluster functionality.
+  * The upgrade process should be carried out in a rolling fashion so that not all components are being updated simultaneously.
+The cluster should be maintained in a healthy state the entire time.
+* **Migrations:** Breaking changes, as well as schema and data format changes should be handled through an automated migration processes.
+* **Rollback:** In the event that the upgrade is not successful, the Rook software should be rolled back to the previous version and cluster health should be restored.
+
+## **User Guide**
+Until automated upgrade support is available in Rook, we have authored a user guide that walks you through the steps to upgrade the software in a Rook cluster.
+Consideration is also provided in the guide for how to verify the cluster remains healthy during and after the upgrade process.
+Please refer to the [Rook Upgrade User Guide](../Documentation/upgrade.md) to learn more about the current Rook upgrade process.
+
+## **Detailed Design**
+The responsibility for performing and orchestrating an upgrade will be handled by an upgrade controller that runs as part of the Rook operator, in the same pod and process (similar to how the Rook volume provisioner is run).
+This controller will be responsible for carrying out the sequence of steps for updating each individual Rook component.
+Additionally, the controller will monitor cluster and component health during the upgrade process, taking corrective steps to restore health, up to and including a full rollback to the old version.
+
+### **Prerequisites**
+In order for the upgrade controller to begin an upgrade process, the following conditions must be met:
+* The cluster should be in a healthy state in accordance with our defined [health verification checks](#upgrade-health-verification).
+The upgrade controller should not begin an upgrade if the cluster is currently unhealthy.
+* Metadata for pods must be persistent.  If config files and other metadata only resides on an ephemeral empty dir for the pods (i.e., `dataDirHostPath` is not set), then the upgrade controller will not perform an upgrade.
+
+### **General Sequence**
+This section describes in a broad sense the general sequence of steps for upgrading a Rook cluster.
+Note that this sequence is modeled after the [Rook Upgrade User Guide](../Documentation/upgrade.md), including the cluster health checks described in the [health verification section](../Documentation/upgrade.md#health-verification).
```

Comment:
something to add to the sequence is that we should store the state of the upgrade as it progresses in the CRD. We aren't yet using the state, but need to for areas like the upgrade. This would allow the upgrade to recover if it gets interrupted in the middle.

Reply:
good point, i'll add a note of keeping track of the upgrade progress in the status of the cluster CRD.